thunx-pdp-opa/src/main/java/com/contentgrid/thunx/pdp/opa/QuerySetToThunkExpressionConverter.java

```diff
@@ -279,6 +287,89 @@ public ThunkExpression<?> visit(Null nullValue) {
 
         @Override
         public ThunkExpression<?> visit(ArrayTerm arrayTerm) {
+            ScalarCollectingRegoVisitor collectingRegoVisitor = new ScalarCollectingRegoVisitor();
+            arrayTerm.getValue().forEach(scalarTerm -> scalarTerm.accept(collectingRegoVisitor));
+            return Scalar.of(collectingRegoVisitor.getScalars());
+        }
+
+        @Override
+        public ThunkExpression<?> visit(SetTerm setTerm) {
+            ScalarCollectingRegoVisitor collectingRegoVisitor = new ScalarCollectingRegoVisitor();
+            setTerm.getValue().forEach(scalarTerm -> scalarTerm.accept(collectingRegoVisitor));
+            return Scalar.of(collectingRegoVisitor.getScalars());
+        }
+    }
+
+    static class ScalarCollectingRegoVisitor implements RegoVisitor<ThunkExpression<?>> {
+
+        @Getter
+        List<Scalar<?>> scalars = new ArrayList<>();
+
+        @Override
+        public ThunkExpression<?> visit(QuerySet querySet) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Query query) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Expression expression) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Ref ref) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Call call) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Var var) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Numeric numeric) {
+            Scalar<?> scalar = Scalar.of(numeric.getValue());
+            scalars.add(scalar);
+            return scalar;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Text text) {
+            Scalar<?> scalar = Scalar.of(text.getValue());
+            scalars.add(scalar);
+            return scalar;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Bool bool) {
+            return null;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(Null aNull) {
+            Scalar<?> scalar = Scalar.nullValue();
+            scalars.add(scalar);
+            return scalar;
+        }
+
+        @Override
+        public ThunkExpression<?> visit(ArrayTerm arrayTerm) {
+            arrayTerm.getValue().forEach(scalarTerm -> scalarTerm.accept(this));
```

Comment:
Doing this would unpack a rego array `['a', 'b', ['c', 'd']]` into `['a', 'b', 'c', 'd']`, which is not a valid conversion.

Instead, this would probably need to use the same code as used above to collect scalars, and add the whole resulting `CollectionValue` as a scalar value.

Reply:
Now SetTerm and ArrayTerm extends Term, that means that SetTerm cannot have SetTerm as a value. If you want to have such case when SetTerm contains SetTerm, then we need to change opa-client, to make SetTerm and ArrayTerm extend ScalarTerm

Reply:
I've added support of collection in collection for opa-client (https://github.com/xenit-eu/opa-java-client/pull/28) and for thunx